docs/resources/server/damagetypes.md

```diff
@@ -0,0 +1,127 @@
+        null,
+        // The direct entity. For example, if a skeleton shot you, the skeleton would be the causing entity
+        // (= the parameter above), and the arrow would be the direct entity (= this parameter). Similar to
+        // the causing entity, this isn't always applicable and therefore nullable. Optional, defaults to null.
+        null,
+        // The damage source position. This is rarely used, one example would be intentional game design
+        // (= nether beds exploding). Nullable and optional, defaulting to null.
+        null
+);
+```
+
+If `DamageSource`s have no entity or position context whatsoever, it makes sense to cache them in a field. For `DamageSource`s that do have entity or position context, it is common to add helper methods, like so:
+
+```java
+public static DamageSource exampleDamage(Entity causer) {
+    return new DamageSource(
+            causer.level().registryAccess().registryOrThrow(Registries.DAMAGE_TYPE).getHolderOrThrow(EXAMPLE_DAMAGE),
+            causer);
+}
+```
+
+:::tip
+Vanilla's `DamageSource` factories can be found in `DamageSources`, and vanilla's `DamageType` resource keys can be found in `DamageTypes`.
+:::
+
+The first and foremost use case for damage sources is `Entity#hurt`. This method is called whenever an entity is receiving damage. To hurt an entity with our own damage type, we simply call `Entity#hurt` ourselves:
+
+```java
+// The second parameter is the amount of damage, in half hearts.
+entity.hurt(exampleDamage(player), 10);
+```
+
+Other damage type-specific behavior, such as invulnerability checks, is often run through damage type [tags]. These are both added by Minecraft and NeoForge and can be found under `DamageTypeTags` and `Tags.DamageTypes`, respectively.
+
+## Datagen
+
+_For more info, see [Data Generation for Datapack Registries][drdatagen]._
+
+Damage type JSON files can be [datagenned][datagen]. To do so, extend `DatapackBuiltinEntriesProvider` and add your damage types in the `RegistrySetBuilder` that is passed to the superconstructor:
+
+```java
+public class MyDamageTypeProvider extends DatapackBuiltinEntriesProvider {
+    public MyDamageTypeProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider) {
+        super(output, lookupProvider, new RegistrySetBuilder().add(Registries.DAMAGE_TYPE, bootstrap -> {
+            // Use new DamageType() to create an in-code representation of a damage type.
+            // The parameters map to the values of the JSON file, in the order seen above.
+            // All parameters except for the message id and the exhaustion value are optional.
+            bootstrap.register(EXAMPLE_DAMAGE, new DamageType(EXAMPLE_DAMAGE.location(),
+                    DamageScaling.WHEN_CAUSED_BY_LIVING_NON_PLAYER,
+                    0.1f,
+                    DamageEffects.HURT,
+                    DeathMessageType.DEFAULT))
+        }), Set.of(ExampleMod.MOD_ID));
```

Comment:
This should be by the JSON since they are the same.

Reply:
Resolved via Discord discussion (second-to-last paragraph of linked message): https://discord.com/channels/313125603924639766/1185208532014997524/1256350727656439841